controlpanel/api/aws.py

```diff
@@ -24,248 +22,415 @@ def iam_arn(resource, account=settings.AWS_ACCOUNT_ID):
+        'TargetBucket': settings.LOGS_BUCKET_NAME,
+        'TargetPrefix': f"{bucket_name}/",
+    }})
+    bucket.meta.client.put_bucket_encryption(
+        Bucket=bucket_name,
+        ServerSideEncryptionConfiguration={'Rules': [{
+            'ApplyServerSideEncryptionByDefault': {
+                'SSEAlgorithm': 'AES256',
+            },
+        }]},
+    )
+    bucket.meta.client.put_public_access_block(
+        Bucket=bucket_name,
+        PublicAccessBlockConfiguration={
+            'BlockPublicAcls': True,
+            'IgnorePublicAcls': True,
+            'BlockPublicPolicy': True,
+            'RestrictPublicBuckets': True,
+        },
+    )
+    return bucket
+
+
+class S3AccessPolicy:
+    """Provides a convenience wrapper around a RolePolicy object"""
+
+    def __init__(self, policy):
+        self.policy = policy
+        self._policy_document = None
+        self.statements = {}
+
+        self._load()
+
+    def _load(self):
         try:
-            result = self._do("iam", f"get_{resource_type}_policy", **aws_kwargs)
-            return result["PolicyDocument"]
-        except ClientError as e:
-            if e.response["Error"]["Code"] == "NoSuchEntity":
-                pass
-            else:
-                raise e
-
-    def get_inline_policy_document(self, role_name, policy_name):
-        return self._get_inline_policy_document(role_name, "role", policy_name)
-
-    def put_role_policy(self, role_name, policy_name, policy_document):
-        self._do(
-            "iam",
-            "put_role_policy",
-            RoleName=role_name,
-            PolicyName=policy_name,
+            # ensure version is set
+            self.policy_document['Version'] = '2012-10-17'
+
+        except self.policy.meta.client.exceptions.NoSuchEntityException:
+            # create an empty s3 access policy
+            self.put(policy_document=dict(BASE_S3_ACCESS_POLICY))
+
+        # ensure statements are correctly configured and build a lookup table
+        for stmt in self.policy_document.setdefault('Statement', []):
+            sid = stmt.get('Sid')
+            if sid in ('list', 'readonly', 'readwrite'):
+                stmt.update(BASE_S3_ACCESS_STATEMENT[sid])
+                self.statements[sid] = stmt
+
+    @property
+    def policy_document(self):
+        if self._policy_document is None:
+            self._policy_document = self.policy.policy_document
+        return self._policy_document
+
+    def statement(self, sid):
+        if sid in ('list', 'readonly', 'readwrite'):
+            if sid not in self.statements:
+                stmt = dict(BASE_S3_ACCESS_STATEMENT[sid])
+                self.statements[sid] = stmt
+                self.policy_document['Statement'].append(stmt)
+            return self.statements[sid]
+
+    def add_resource(self, arn, sid):
+        statement = self.statement(sid)
+        if statement:
+            resources = statement.setdefault('Resource', [])
```

Comment:
This is weird, so this is to tell the dictionary to return an empty list when the key is not present right?

Wouldn't be easier to just add the `"Resource": []` key to the statements in the "constants"?

Reply:
`setdefault` saves a line. Without it we have to check if the statement dict has a `Resource` key and if it is missing (possible, but unlikely), create it, append the arn, then update the statement dict with the newly-created list. Eg:
```python
if statement:
    resources = statement.get('Resource', [])
    if arn not in resources:
        resources.append(arn)
        statement['Resource'] = resources
```

Reply:
actually, i think i'll do it without `setdefault`